File: mobile/src/main/java/treehou/se/habit/ui/homescreen/VoiceService.java
package treehou.se.habit.ui.homescreen;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.speech.RecognizerIntent;
import android.util.Log;

import java.util.List;

import io.realm.Realm;
import se.treehou.ng.ohcommunicator.services.Connector;
import se.treehou.ng.ohcommunicator.services.IServerHandler;
import treehou.se.habit.core.db.model.ServerDB;

public class VoiceService extends IntentService {

    public static final String TAG = "VoiceService";

    private static final String ACTION_COMMAND = "treehou.se.habit.ui.homescreen.action.VOICE";
    public static final String EXTRA_SERVER = "extraServerId";

    public static final String VOICE_ITEM = "VoiceCommand";

    private static final int NULL_SERVER = -1;

    public static Intent createVoiceCommand(Context context, ServerDB server) {
        Intent intent = new Intent(context, VoiceService.class);
        intent.setAction(ACTION_COMMAND);
        intent.putExtra(EXTRA_SERVER, server.getId());
        return intent;
    }

    public VoiceService() {
        super("VoiceService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        Log.w(TAG, "onHandleIntent.");

        long serverId = intent.getLongExtra(EXTRA_SERVER, NULL_SERVER);
        if(NULL_SERVER == serverId){
            Log.w(TAG, "No server specified.");
            return;
        }

        Realm realm = Realm.getDefaultInstance();
        ServerDB server = ServerDB.load(realm, serverId);

        List<String> results = intent.getExtras().getStringArrayList(RecognizerIntent.EXTRA_RESULTS);
        if (results != null && !results.isEmpty() && server != null) {
            Log.d(TAG, "Received " + results.size() + " voice results.");

            String command = results.get(0);
            IServerHandler serverHandler = new Connector.ServerHandler(server.toGeneric(), this);
            serverHandler.sendCommand(VOICE_ITEM, command);
        }
        realm.close();
    }
}

File: mobile/src/main/java/treehou/se/habit/ui/homescreen/VoiceService.kt
package treehou.se.habit.ui.homescreen

import android.app.IntentService
import android.content.Context
import android.content.Intent
import android.speech.RecognizerIntent
import android.support.v4.app.JobIntentService
import android.util.Log

import javax.inject.Inject

import io.realm.Realm
import treehou.se.habit.core.db.model.ServerDB
import treehou.se.habit.util.ConnectionFactory
import treehou.se.habit.util.Util

class VoiceService : JobIntentService() {

    @Inject lateinit var connectionFactory: ConnectionFactory

    override fun onCreate() {
        super.onCreate()
        Util.getApplicationComponent(this).inject(this)
    }

    override fun onHandleWork(intent: Intent) {
        Log.w(TAG, "onHandleIntent.")

        val serverId = intent.getLongExtra(EXTRA_SERVER, NULL_SERVER.toLong())
        if (NULL_SERVER.toLong() == serverId) {
            Log.w(TAG, "No server specified.")
            return
        }

        val realm = Realm.getDefaultInstance()
        val server = ServerDB.load(realm, serverId)

        val results = intent.extras!!.getStringArrayList(RecognizerIntent.EXTRA_RESULTS)
        if (results != null && !results.isEmpty() && server != null) {
            Log.d(TAG, "Received " + results.size + " voice results.")

            val command = results[0]
            val serverHandler = connectionFactory.createServerHandler(server.toGeneric(), this)
            serverHandler.sendCommand(VOICE_ITEM, command)
        }
        realm.close()
    }

    companion object {

        val TAG = "VoiceService"

        private val ACTION_COMMAND = "treehou.se.habit.ui.homescreen.action.VOICE"
        val EXTRA_SERVER = "extraServerId"

        val VOICE_ITEM = "VoiceCommand"

        private val NULL_SERVER = -1

        fun createVoiceCommand(context: Context, server: ServerDB): Intent {
            val intent = Intent(context, VoiceService::class.java)
            intent.action = ACTION_COMMAND
            intent.putExtra(EXTRA_SERVER, server.id)
            return intent
        }
    }
}
